Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE use in TextCodecUTF16::decode()

cdumez opened the pull request:
#### 4d5eae2982107109e9a03c38de31eee755195687
<pre>
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE use in TextCodecUTF16::decode()
<a href="https://bugs.webkit.org/show_bug.cgi?id=285387">https://bugs.webkit.org/show_bug.cgi?id=285387</a>

Reviewed by Darin Adler.

* Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
(PAL::TextCodecUTF16::decode):

Canonical link: <a href="https://commits.webkit.org/288454@main">https://commits.webkit.org/288454@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4693efe5e065e621fdea1b7c28cfb62124483801

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3070 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88529 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34463 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85541 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3155 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11027 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/64921 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22663 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86503 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2303 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75828 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45211 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2206 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30034 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33513 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30770 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89905 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10718 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7725 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/73346 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10943 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71655 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72575 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16807 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15526 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/2051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12877 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10673 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/16142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10524 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13995 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12295 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
```diff
@@ -66,14 +66,11 @@ void TextCodecUTF16::registerCodecs(TextCodecRegistrar registrar)
-    const auto* const end = p + bytes.size();
-    const auto* const endMinusOneOrNull = end ? end - 1 : nullptr;
+    size_t index = 0;
+    const size_t lengthMinusOne = bytes.size() - 1;
```
```suggestion
    size_t lengthMinusOne = bytes.size() - 1;
```
So many local variables _could_ be marked `const` but we generally don't do it, for brevity's sake. Although I see that the old code did.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
```diff
@@ -109,32 +106,34 @@ String TextCodecUTF16::decode(std::span<const uint8_t> bytes, bool flush, bool,
-            processBytesBE(p[0], p[1]);
-            p += 2;
+
+        if (m_littleEndian) {
```
```suggestion
        if (m_littleEndian) {
```

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
```diff
@@ -109,32 +106,34 @@ String TextCodecUTF16::decode(std::span<const uint8_t> bytes, bool flush, bool,
+                processBytesLE(leadByte, bytes[index]);
+            else
+                processBytesBE(leadByte, bytes[index]);
+            ++index;
```
```suggestion
            auto leadByte = *std::exchange(m_leadByte, std::nullopt);
            auto trailByte = bytes[index++];
            if (m_littleEndian)
                processBytesLE(leadByte, trailByte);
            else
                processBytesBE(leadByte, trailByte);
```

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
```diff
@@ -109,32 +106,34 @@ String TextCodecUTF16::decode(std::span<const uint8_t> bytes, bool flush, bool,
+            while (index < lengthMinusOne) {
+                processBytesLE(bytes[index], bytes[index + 1]);
+                index += 2;
+            }
```
```suggestion
            for (; index < lengthMinusOne; index += 2)
                processBytesLE(bytes[index], bytes[index + 1]);
```

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
```diff
@@ -109,32 +106,34 @@ String TextCodecUTF16::decode(std::span<const uint8_t> bytes, bool flush, bool,
+            while (index < lengthMinusOne) {
+                processBytesBE(bytes[index], bytes[index + 1]);
+                index += 2;
+            }
```
```suggestion
            for (; index < lengthMinusOne; index += 2)
                processBytesBE(bytes[index], bytes[index + 1]);
```

cdumez committed:
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE use in TextCodecUTF16::decode()
https://bugs.webkit.org/show_bug.cgi?id=285387

Reviewed by Darin Adler.

* Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
(PAL::TextCodecUTF16::decode):

Canonical link: https://commits.webkit.org/288454@main
